Index: lang/en/editor_tinymce.php
===================================================================
RCS file: /cvsroot/moodle/moodle/lang/en/editor_tinymce.php,v
retrieving revision 1.7
diff -u -r1.7 editor_tinymce.php
--- lang/en/editor_tinymce.php	19 Jul 2009 13:05:24 -0000	1.7
+++ lang/en/editor_tinymce.php	10 Nov 2009 08:09:43 -0000
@@ -234,6 +234,8 @@
 $string['autosave:unload_msg'] = 'The changes you made will be lost if you navigate away from this page.';
 $string['common:apply'] = 'Apply';
 $string['common:browse'] = 'Browse';
+$string['common:browsemedia'] = 'Find or upload a sound, video or applet';
+$string['common:browseimage'] = 'Find or upload an image';
 $string['common:cancel'] = 'Cancel';
 $string['common:class_name'] = 'Class';
 $string['common:clipboard_msg'] = 'Copy/Cut/Paste is not available in Mozilla and Firefox.\\nDo you want more information about this issue?';
@@ -414,6 +416,7 @@
 $string['media_dlg:enablejavascript'] = 'Enable JavaScript';
 $string['media_dlg:endtime'] = 'End time';
 $string['media_dlg:file'] = 'File/URL';
+$string['media_dlg:filename'] = 'Filename';
 $string['media_dlg:flash_options'] = 'Flash options';
 $string['media_dlg:flashvars'] = 'Flashvars';
 $string['media_dlg:flv_autostart'] = 'Auto start';
