Index: tiny_mce_src.js
===================================================================
RCS file: /cvsroot/moodle/moodle/lib/editor/tinymce/tiny_mce/3.3.8/tiny_mce_src.js,v
retrieving revision 1.1
diff -u -r1.1 tiny_mce_src.js
--- tiny_mce_src.js 4 Jul 2010 19:21:59 -0000   1.1
+++ tiny_mce_src.js 4 Jul 2010 20:11:25 -0000
@@ -8744,6 +8744,7 @@
        },
 
        requireLangPack : function(n) {
+           return; // not needed in Moodle - we load all lang packs in one request
            var s = tinymce.settings;
 
            if (s && s.language)
@@ -9334,7 +9335,8 @@
            // Load scripts
            function loadScripts() {
                if (s.language)
-                   sl.add(tinymce.baseURL + '/langs/' + s.language + '.js');
+                   sl.add(tinymce.baseURL + '/../../extra/strings.php?elanguage=' + s.language + '&etheme=' + s.theme); // Moodle modification - load one huge strings file
+                   //sl.add(tinymce.baseURL + '/langs/' + s.language + '.js');
 
                if (s.theme && s.theme.charAt(0) != '-' && !ThemeManager.urls[s.theme])
                    ThemeManager.load(s.theme, 'themes/' + s.theme + '/editor_template' + tinymce.suffix + '.js');
Index: Popup.js
===================================================================
RCS file: /cvsroot/moodle/moodle/lib/editor/tinymce/tiny_mce/3.3.8/Popup.js,v
retrieving revision 1.1
diff -u -r1.1 Popup.js
--- Popup.js    4 Jul 2010 19:21:59 -0000   1.1
+++ Popup.js    4 Jul 2010 20:11:19 -0000
@@ -187,6 +187,8 @@
     * @method requireLangPack
     */
    requireLangPack : function() {
+       return; // not needed in Moodle - we load all lang packs in one request
+
        var t = this, u = t.getWindowArg('plugin_url') || t.getWindowArg('theme_url');
 
        if (u && t.editor.settings.language && t.features.translate_i18n !== false) {
