Index: lib/editor/tinymce/plugins/media/js/media.js
===================================================================
RCS file: /cvsroot/moodle/moodle/lib/editor/tinymce/plugins/media/js/media.js,v
retrieving revision 1.1
diff -u -r1.1 media.js
--- lib/editor/tinymce/plugins/media/js/media.js    16 May 2009 20:54:41 -0000  1.1
+++ lib/editor/tinymce/plugins/media/js/media.js    17 May 2009 07:44:27 -0000
@@ -263,6 +263,13 @@
 
        h += ' />';
 
+        var is_youtube = repo_hook(f.src.value);
+        if (is_youtube) {
+            var re = new RegExp("v\/(.+)\#(.+)", "i");
+            var result = f.src.value.match(re);
+            h = '<a href="'+f.src.value+'">'+result[2]+'</a>';
+        }
+
        ed.execCommand('mceInsertContent', false, h);
    }
 
@@ -310,7 +317,11 @@
        f.height.value = '350';
        f.src.value = 'http://www.youtube.com/v/' + v.match(/v=(.*)(.*)/)[0].split('=')[1];
        return 'flash';
-   }
+   } else if (v.match(/v\/(.+)(.*)/)) {
+       f.width.value = '425';
+       f.height.value = '350';
+       return 'flash';
+    }
 
    // Google video
    if (v.indexOf('http://video.google.com/videoplay?docid=') == 0) {
@@ -627,4 +638,15 @@
    p.innerHTML = "<!-- x --->" + h;
 }
 
+function repo_hook(url){
+   var f = document.forms[0];
+   if (url.match(/v\/(.+)(.*)/)) {
+       f.width.value = '425';
+       f.height.value = '350';
+        return true;
+   } else {
+        return false;
+    }
+}
+
 tinyMCEPopup.onInit.add(init);
